code/modules/station_goals/bsa.dm

```diff
@@ -178,17 +178,42 @@
 	reload()
 
 /obj/machinery/bsa/full/proc/fire(mob/user, turf/bullseye)
+	reload()
+
 	var/turf/point = get_front_turf()
-	for(var/turf/T in getline(get_step(point,dir),get_target_turf()))
-		T.ex_act(EXPLODE_DEVASTATE)
-	point.Beam(get_target_turf(),icon_state="bsa_beam",time=50,maxdistance = world.maxx) //ZZZAP
+	var/turf/target = get_target_turf()
+	var/targ_x = 1
+	var/targ_y = 1
+	var/atom/movable/blocker
+	for(var/T in getline(get_step(point, dir), target))
+		var/turf/tile = T
+		targ_x = tile.x
+		targ_y = tile.y
+		if(SEND_SIGNAL(tile, COMSIG_ATOM_BSA_BEAM) & COMSIG_ATOM_BLOCKS_BSA_BEAM)
+			blocker = tile
+		else
+			for(var/AM in tile)
+				var/atom/movable/stuff = AM
+				if(SEND_SIGNAL(stuff, COMSIG_ATOM_BSA_BEAM) & COMSIG_ATOM_BLOCKS_BSA_BEAM)
+					blocker = stuff
+					break
+		if(blocker) //BYOND's only mechanism for escaping from an inner foreach loop is GOTO. If I did that, this code would GOTO the trash can
+			target = locate(targ_x, targ_y, z) //both get_turf(blocker) and T may be deleted by this point
```

Comment:
Turf references don't null and just refer to the new turf

Reply:
I always forget how turfs worked, I'm pretty sure I've literally said this to other people before.